storage/pin/pin_test.go

```diff
@@ -0,0 +1,328 @@
+	if err != nil {
+		t.Fatalf("could not create new manifest error: %v", err.Error())
+	}
+
+	newAddr, err = p.api.UpdateManifest(context.TODO(), newAddr, func(mw *api.ManifestWriter) error {
+
+		for _, fileInfo := range testDirFiles {
+			entry := &api.ManifestEntry{
+				Hash:        hex.EncodeToString(fileInfo.fileHash),
+				Path:        fileInfo.fileName,
+				ContentType: mime.TypeByExtension(filepath.Ext(fileInfo.fileName)),
+			}
+			_, err = mw.AddEntry(context.TODO(), nil, entry)
+			if err != nil {
+				t.Fatalf("could not create new manifest error: %v", err.Error())
+			}
+		}
+		return nil
+	})
+
+	return newAddr
+}
+
+// This function is called from test after a file is pinned.
+// It check if the file's chunks are properly pinned.
+// Assumption is that the file is uploaded in an empty database so that it can be easily tested.
+// It takes the root hash and expected pinCounter value.
+// It also has some hacks to take care of existing issues in the way we upload.
+//
+// The check process is as follows
+//   1) Check if the root hash is pinned in pinFilesIndex
+//         - getPinnedFiles gives all the root chunks pinned in pinFilesIndex
+//         - Check if this list contains our root hash
+//   2) Check if all the files's chunks are in pinIndex
+//         a) Get all the chunks
+//            get it from retrievalDataIndex
+//            since the assumption is the DB has only this file, it gives all the file's chunks.
+//            getAllRefs cannot be used here as it does not give the chunks that belong to manifests.
+//         b) Get all chunks that are pinned (from pinIndex)
+//            In every upload.. an empty manifest is uploaded. that why add this hash to this list
+//         c) Check if both the above lists are equal
+//   3) Check if all the chunks pinned have the proper pinCounter
+//         -  This is just a simple go through of all the pinned chunks list and check if the counter is
+//            equal to the pin counter given as argument
+func failIfNotPinned(t *testing.T, p *PinAPI, rootHash string, pinCounter uint64, isRaw bool) {
+
+	t.Helper()
+
+	// 1 - Check if the root hash is pinned in pinFilesIndex
+	pinnedFiles := p.getPinnedFiles()
+	if _, ok := pinnedFiles[rootHash]; !ok {
+		t.Fatalf("File %s not pinned in pinFilesIndex", rootHash)
+	}
+
+	// 2a - Get all the chunks of the file from retrievalDataIndex (since the file is uploaded in an empty database)
+	chunksInDB := p.getAllChunksFromDB()
+
+	// 2b - Get pinned chunks details from pinning indexes
+	pinnedChunks := p.collectPinnedChunks(rootHash, "")
+	if !isRaw {
+		// Add the empty manifest chunk
+		pinnedChunks["8b634aea26eec353ac0ecbec20c94f44d6f8d11f38d4578a4c207a84c74ef731"] = pinCounter
+	}
+
+	// 2c - Check if number of chunk hashes are same
+	if len(chunksInDB) != len(pinnedChunks) {
+		t.Fatalf("Expected number of chunks to be %d, but is %d", len(chunksInDB), len(pinnedChunks))
+	}
+
+	// 2c - Check if all the chunk address are same
```

Comment:
duplicate comment

Reply:
Nope.. 2c has two checks... one checks the length and the other check all the contents.